File: HomeWork 2/tree.cpp
//------------------------------------------------------------------------------
// tree.cpp - содержит функции обработки дерева
//------------------------------------------------------------------------------

#include "tree.h"

//------------------------------------------------------------------------------
// Ввод параметров дерева из потока
void tree::In(std::ifstream &ifst) {
    ifst >> this->age >> this->name;
}

// Случайный ввод параметров дерева
void tree::InRnd() {
    this->age = Random();
    for (size_t i = 0; i < MAX_NAME_SIZE; ++i) {
        this->name[i] = (char)('a' + Random());
    }
    this->name[MAX_NAME_SIZE - 1] = '\0';
}

//------------------------------------------------------------------------------
// Вывод параметров дерева в поток
void tree::Out(std::ofstream &ofst) {
    ofst << "It is tree: age = "
         << this->age << ", name = " << this->name
         << ". CountExpr = " << this->CountExpr() << "\n";
}

//------------------------------------------------------------------------------
// Вычисление частного от деления числа гласных букв
// в названии дерева на общую длину его названия

double tree::CountExpr() {
    // переменная для подсчёта гласных в названии дерева
    size_t number_of_vowels = 0;
    for (auto it: this->name) {
        char current_symbol = tolower(it);
        if (current_symbol == 'a' || current_symbol == 'e' || current_symbol == 'i' ||
            current_symbol == 'o' || current_symbol == 'u' || current_symbol == 'y') {
            ++number_of_vowels;
        }
    }
    return (double)number_of_vowels / strlen(this->name);
}

tree::~tree() {

}

File: HomeWork 3/tree.py
from rnd import Random, randName

# Класс дерева
class Tree :
    # Конструктор дерева
    def __init__(self):
        self.name = ""
        self.age = 0
    
    # Считывание параметров дерева из строки
    def ReadStrArray(self, strArray, i):
        if i >= len(strArray) - 1:
            return 0
        self.name = strArray[i]
        self.age = int(strArray[i + 1])
        i += 2
        return i

    # Случайная генерация параметров дерева
    def RandomIn(self):
        self.name = randName()
        self.age = Random()

    # Вывод данных дерева в файл
    def Write(self, ostream):
        ostream.write(f"Tree: name = {self.name}, age = {self.age}, CountExpr = {self.CountExpr()}")

    # Вычисление частного от деления числа гласных букв
    # в названии дерева на общую длину его названия
    def CountExpr(self):
        numberOfVowels = 0
        for ch in self.name:
            if(ch == 'a' or ch == 'e' or ch == 'i' or ch == 'o' or ch == 'u' or ch == 'A'
                or ch == 'E' or ch == 'I' or ch == 'O' or ch == 'U'):
                numberOfVowels += 1
        return numberOfVowels / len(self.name)